nanne-aben/strictly_typed_pandas
# Vendor typeguard 2.13.3

Vendor the old version of typeguard that is used (2.13.3), so that projects are free to upgrade or not install typeguard as needed (eg for other packages that depend on newer versions).

I have tried to make this as smooth as possible, to require project maintainers to make as few changes as possible and to allow it to work in as many environments as possible. To do that, I have edited the pytest plugin to allow projects to continue to use `--typeguard-packages` if typeguard is not installed, or if a compatible version of typeguard is installed.

If an incompatible version of typeguard is installed, project maintainers will need to use `--stp-typeguard-packages`.

## How to test
```bash
pip install -e .
pip uninstall typeguard
pytest --typeguard-packages=tests
pytest --stp-typeguard-packages=tests
pytest --typeguard-packages=tests --stp-typeguard-packages=tests  # expect exception

pip install typeguard==2.13.3
pytest --typeguard-packages=tests
pytest --stp-typeguard-packages=tests

pip install typeguard==4.1.5
pytest --stp-typeguard-packages=tests
pytest --typeguard-packages=tests  # expect fail
```

Double check the typeguard code is correct:

```
pip install python-vendorize
rm -fr strictly_typed_pandas/_vendor/*
python-vendorize
git diff  # There were some minor changes made, inspect them
```


## To do...
- [x] Check if a compatible version of typeguard is installed, and to allow `--typeguard-packages` to be checked by that version.
- [x] I've tried to exclude the vendored code from pre-commit, but when it runs for "all files" the vendored code is checked. I tried to add configuration for each of the tools, but it wasn't always picked up. Updating the tool configuration is a better idea, because it will also work when the tools (eg mypy) are run manually (ie not via pre-commit).
- [x] Allowing `--typeguard-packages` to be used is a convenience, but it will silently fail in the following situation. We should try to identify that situation, and raise an exception:
   - end developer is using `strictly_typed_pandas` and wants to make use of `typeguard`
   - has a newer (unsupported) version of `typeguard` installed (eg 3+)
   - uses `--typeguard-packages` for those packages where they should be using `--stp-typeguard-packages`
   - UPDATE: I don't think this is possible, so I'll recommend only the --stp-typeguard-packages option in the documentation and hope for the best. We could alternatively output a python warning that the user can ignore if they know what they're doing.
